File: android/app/src/main/java/com/gamestream/app/CloudCatalogService.kt
package com.gamestream.app

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

object CloudCatalogService {
    private const val SIGL =
        "https://catalog.gamepass.com/sigls/v2?id=29a81209-df6f-41fd-a528-2ae6b91f719c&language=en-us&market=US"
    private const val CACHE = "xcloud-catalog-v1.json"
    @Volatile private var started = false

    fun refreshIfNeeded(context: Context) {
        if (started) return
        started = true
        loadCache(context)?.takeIf { it.size >= 20 }?.let { GameCatalog.installLiveCatalog(it) }
    }

    suspend fun fetchAndInstall(context: Context) = withContext(Dispatchers.IO) {
        refreshIfNeeded(context)
        try {
            val remote = fetchRemote()
            if (remote.size >= 20) {
                saveCache(context, remote)
                withContext(Dispatchers.Main) {
                    GameCatalog.installLiveCatalog(remote)
                }
            }
        } catch (_: Exception) {
        }
    }

    private fun fetchRemote(): List<CatalogGame> {
        val raw = JSONArray(httpGet(SIGL))
        val ids = linkedSetOf<String>()
        for (i in 0 until raw.length()) {
            val row = raw.optJSONObject(i) ?: continue
            val id = row.optString("id").trim()
            if (id.isNotEmpty()) ids.add(id)
        }
        val games = mutableListOf<CatalogGame>()
        val list = ids.toList()
        var index = 0
        while (index < list.size) {
            val slice = list.subList(index, minOf(index + 20, list.size))
            index += 20
            games += hydrate(slice)
        }
        return games.distinctBy { it.id.uppercase() }
    }

    private fun hydrate(ids: List<String>): List<CatalogGame> {
        val joined = ids.joinToString(",")
        val url =
            "https://displaycatalog.mp.microsoft.com/v7.0/products?bigIds=$joined&market=US&languages=en-us&MS-CV=GS.1"
        val json = JSONObject(httpGet(url))
        val products = json.optJSONArray("Products") ?: return emptyList()
        val out = mutableListOf<CatalogGame>()
        for (i in 0 until products.length()) {
            val product = products.optJSONObject(i) ?: continue
            val id = product.optString("ProductId").trim()
            if (id.isEmpty()) continue
            val loc = product.optJSONArray("LocalizedProperties")?.optJSONObject(0)
            val title = loc?.optString("ProductTitle")?.trim().orEmpty().ifEmpty { id }
            val tagline = loc?.optString("ShortDescription")?.trim().orEmpty().take(140)
            val props = product.optJSONObject("Properties")
            val genre = props?.optString("Category")?.takeIf { it.isNotBlank() }
                ?: props?.optJSONArray("Categories")?.optString(0)
                ?: "Cloud"
            val images = loc?.optJSONArray("Images")
            out += CatalogGame(
                id = id,
                slug = slugify(title),
                title = title,
                tagline = tagline,
                genre = shortenGenre(genre),
                featured = false,
                accent = 0xFF4361EE,
                posterUrl = pickPoster(images)
            )
        }
        return out
    }

    private fun pickPoster(images: JSONArray?): String? {
        if (images == null) return null
        val preferred = listOf("Poster", "BoxArt", "BrandedKeyArt", "TitledHeroArt", "SuperHeroArt")
        fun normalize(raw: String): String? {
            if (raw.isBlank()) return null
            return if (raw.startsWith("//")) "https:$raw" else raw
        }
        for (purpose in preferred) {
            for (i in 0 until images.length()) {
                val img = images.optJSONObject(i) ?: continue
                if (img.optString("ImagePurpose") == purpose) {
                    normalize(img.optString("Uri"))?.let { return it }
                }
            }
        }
        return normalize(images.optJSONObject(0)?.optString("Uri").orEmpty())
    }

    private fun slugify(title: String): String {
        val out = StringBuilder()
        var dash = false
        for (ch in title.lowercase()) {
            if (ch.isLetterOrDigit()) {
                out.append(ch)
                dash = false
            } else if (!dash) {
                out.append('-')
                dash = true
            }
        }
        return out.trim('-').toString()
    }

    private fun shortenGenre(raw: String): String {
        val value = raw.lowercase()
        return when {
            "race" in value -> "Racing"
            "shoot" in value -> "Shooter"
            "rpg" in value || "role" in value -> "RPG"
            "platform" in value -> "Platformer"
            "sandbox" in value -> "Sandbox"
            "surviv" in value -> "Survival"
            "sport" in value -> "Sports"
            "fight" in value -> "Fighting"
            "sim" in value -> "Simulation"
            "strat" in value -> "Strategy"
            "puzzle" in value -> "Puzzle"
            "adventure" in value -> "Adventure"
            "action" in value -> "Action"
            else -> raw
        }
    }

    private fun cacheFile(context: Context) = File(context.cacheDir, CACHE)

    private fun saveCache(context: Context, games: List<CatalogGame>) {
        val arr = JSONArray()
        games.forEach { game ->
            arr.put(
                JSONObject()
                    .put("id", game.id)
                    .put("slug", game.slug)
                    .put("title", game.title)
                    .put("tagline", game.tagline)
                    .put("genre", game.genre)
                    .put("poster", game.posterUrl.orEmpty())
            )
        }
        cacheFile(context).writeText(arr.toString())
    }

    private fun loadCache(context: Context): List<CatalogGame>? {
        val file = cacheFile(context)
        if (!file.exists()) return null
        return try {
            val arr = JSONArray(file.readText())
            val seen = mutableSetOf<String>()
            val games = mutableListOf<CatalogGame>()
            for (i in 0 until arr.length()) {
                val row = arr.optJSONObject(i) ?: continue
                val id = row.optString("id")
                if (id.isBlank() || !seen.add(id.uppercase())) continue
                val poster = row.optString("poster").ifBlank { null }
                games += CatalogGame(
                    id = id,
                    slug = row.optString("slug").ifBlank { id.lowercase() },
                    title = row.optString("title").ifBlank { id },
                    tagline = row.optString("tagline"),
                    genre = row.optString("genre").ifBlank { "Cloud" },
                    posterUrl = poster
                )
            }
            games
        } catch (_: Exception) {
            null
        }
    }

    private fun httpGet(url: String): String {
        val conn = URL(url).openConnection() as HttpURLConnection
        conn.connectTimeout = 15000
        conn.readTimeout = 20000
        conn.setRequestProperty("Accept", "application/json")
        conn.inputStream.bufferedReader().use { return it.readText() }
    }
}

File: GameStream/Services/CloudCatalogService.swift
import Foundation

enum CloudCatalogService {
    private static let siglURL = URL(string: "https://catalog.gamepass.com/sigls/v2?id=29a81209-df6f-41fd-a528-2ae6b91f719c&language=en-us&market=US")!
    private static let cacheName = "xcloud-catalog-v2.json"
    private static var started = false

    static func refreshIfNeeded() {
        guard !started else { return }
        started = true
        Task { @MainActor in
            if let cached = loadCache(), cached.count >= 20 {
                GameCatalog.installLiveCatalog(cached)
            }
        }
        Task.detached(priority: .utility) {
            do {
                try await fetchRemoteProgressive()
            } catch {
            }
        }
    }

    private static func fetchRemoteProgressive() async throws {
        var request = URLRequest(url: siglURL)
        request.timeoutInterval = 15
        request.cachePolicy = .reloadIgnoringLocalCacheData
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
        let ids = parseIds(from: data)
        guard ids.count >= 20 else { return }

        var collected: [CatalogGame] = []
        collected.reserveCapacity(ids.count)
        let chunk = 20
        var index = 0
        var published = false
        while index < ids.count {
            let end = min(index + chunk, ids.count)
            let slice = Array(ids[index..<end])
            index = end
            do {
                collected.append(contentsOf: try await hydrate(slice))
            } catch {
                continue
            }
            if !published && collected.count >= 24 {
                let snapshot = collected
                await MainActor.run {
                    GameCatalog.installLiveCatalog(snapshot)
                }
                published = true
            }
        }
        let unique = dedupe(collected)
        guard unique.count >= 20 else { return }
        saveCache(unique)
        await MainActor.run {
            GameCatalog.installLiveCatalog(unique)
        }
    }

    private static func parseIds(from data: Data) -> [String] {
        var ids: [String] = []
        var seen = Set<String>()
        func take(_ id: String) {
            let key = id.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
            guard !key.isEmpty, seen.insert(key).inserted else { return }
            ids.append(id.trimmingCharacters(in: .whitespacesAndNewlines))
        }
        guard let raw = try? JSONSerialization.jsonObject(with: data) else { return [] }
        if let rows = raw as? [[String: Any]] {
            for row in rows {
                if let id = row["id"] as? String { take(id) }
            }
        } else if let dict = raw as? [String: Any] {
            if let rows = dict["sigls"] as? [[String: Any]] {
                for row in rows {
                    if let id = row["id"] as? String { take(id) }
                }
            }
        }
        return ids
    }

    private static func hydrate(_ ids: [String]) async throws -> [CatalogGame] {
        let joined = ids.joined(separator: ",")
        guard let url = URL(string: "https://displaycatalog.mp.microsoft.com/v7.0/products?bigIds=\(joined)&market=US&languages=en-us&MS-CV=GS.1") else {
            return []
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 15
        request.cachePolicy = .reloadIgnoringLocalCacheData
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
        guard
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let products = json["Products"] as? [[String: Any]]
        else { return [] }

        var out: [CatalogGame] = []
        for product in products {
            guard let id = product["ProductId"] as? String, !id.isEmpty else { continue }
            let loc = (product["LocalizedProperties"] as? [[String: Any]])?.first
            let title = (loc?["ProductTitle"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? id
            let tagline = (loc?["ShortDescription"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let props = product["Properties"] as? [String: Any]
            let genre = (props?["Category"] as? String)
                ?? ((props?["Categories"] as? [String])?.first)
                ?? "Cloud"
            let images = loc?["Images"] as? [[String: Any]] ?? []
            let poster = pickPoster(images)
            out.append(
                CatalogGame(
                    id: id,
                    slug: slugify(title),
                    title: title,
                    tagline: String(tagline.prefix(140)),
                    genre: shortenGenre(genre),
                    provider: "Xbox Cloud",
                    featured: false,
                    accent: 0x4361EE,
                    posterURL: poster
                )
            )
        }
        return out
    }

    private static func pickPoster(_ images: [[String: Any]]) -> URL? {
        let preferred = ["Poster", "BoxArt", "BrandedKeyArt", "TitledHeroArt", "SuperHeroArt"]
        for purpose in preferred {
            if let match = images.first(where: { ($0["ImagePurpose"] as? String) == purpose }),
               let raw = match["Uri"] as? String {
                return normalizedImageURL(raw)
            }
        }
        if let raw = images.first?["Uri"] as? String {
            return normalizedImageURL(raw)
        }
        return nil
    }

    private static func normalizedImageURL(_ raw: String) -> URL? {
        if raw.hasPrefix("//") { return URL(string: "https:" + raw) }
        return URL(string: raw)
    }

    private static func slugify(_ title: String) -> String {
        let lowered = title.lowercased()
        let allowed = CharacterSet.alphanumerics
        var slug = ""
        var dash = false
        for scalar in lowered.unicodeScalars {
            if allowed.contains(scalar) {
                slug.append(Character(scalar))
                dash = false
            } else if !dash {
                slug.append("-")
                dash = true
            }
        }
        return slug.trimmingCharacters(in: CharacterSet(charactersIn: "-"))
    }

    private static func shortenGenre(_ raw: String) -> String {
        let value = raw.lowercased()
        if value.contains("race") { return "Racing" }
        if value.contains("shoot") { return "Shooter" }
        if value.contains("rpg") || value.contains("role") { return "RPG" }
        if value.contains("platform") { return "Platformer" }
        if value.contains("sandbox") { return "Sandbox" }
        if value.contains("surviv") { return "Survival" }
        if value.contains("sport") { return "Sports" }
        if value.contains("fight") { return "Fighting" }
        if value.contains("sim") { return "Simulation" }
        if value.contains("strat") { return "Strategy" }
        if value.contains("puzzle") { return "Puzzle" }
        if value.contains("adventure") { return "Adventure" }
        if value.contains("action") { return "Action" }
        return raw
    }

    private static func dedupe(_ incoming: [CatalogGame]) -> [CatalogGame] {
        var seen = Set<String>()
        var out: [CatalogGame] = []
        for game in incoming {
            let key = game.id.uppercased()
            if seen.insert(key).inserted {
                out.append(game)
            }
        }
        return out
    }

    private static func cacheURL() -> URL? {
        FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?
            .appendingPathComponent(cacheName)
    }

    static func clearDiskCache() {
        guard let url = cacheURL() else { return }
        try? FileManager.default.removeItem(at: url)
    }

    private static func saveCache(_ games: [CatalogGame]) {
        guard let url = cacheURL() else { return }
        let payload: [[String: String]] = games.map { game in
            [
                "id": game.id,
                "slug": game.slug,
                "title": game.title,
                "tagline": game.tagline,
                "genre": game.genre,
                "poster": game.posterURL?.absoluteString ?? ""
            ]
        }
        if let data = try? JSONSerialization.data(withJSONObject: payload) {
            try? data.write(to: url, options: .atomic)
        }
    }

    private static func loadCache() -> [CatalogGame]? {
        guard
            let url = cacheURL(),
            let data = try? Data(contentsOf: url),
            let rows = try? JSONSerialization.jsonObject(with: data) as? [[String: String]]
        else { return nil }
        var seen = Set<String>()
        var games: [CatalogGame] = []
        for row in rows {
            guard let id = row["id"], seen.insert(id.uppercased()).inserted else { continue }
            let poster = row["poster"].flatMap { $0.isEmpty ? nil : URL(string: $0) }
            games.append(
                CatalogGame(
                    id: id,
                    slug: row["slug"] ?? id.lowercased(),
                    title: row["title"] ?? id,
                    tagline: row["tagline"] ?? "",
                    genre: row["genre"] ?? "Cloud",
                    provider: "Xbox Cloud",
                    featured: false,
                    accent: 0x4361EE,
                    posterURL: poster
                )
            )
        }
        return games
    }
}
